Core/src/org/sleuthkit/autopsy/datamodel/utils/AbstractNodePropertySheetTask.java

```diff
@@ -0,0 +1,140 @@
+ * provided for executing instances of the tasks.
+ */
+public abstract class AbstractNodePropertySheetTask implements Runnable {
+
+    private static final Logger LOGGER = Logger.getLogger(AbstractContentNode.class.getName());
+    private static final Integer THREAD_POOL_SIZE = 10;
+    private static final ExecutorService executor = Executors.newFixedThreadPool(THREAD_POOL_SIZE, new ThreadFactoryBuilder().setNameFormat("node-background-task-%d").build());
+    private final WeakReference<AbstractNode> weakNodeRef;
+    private final WeakReference<PropertyChangeListener> weakListenerRef;
+
+    /**
+     * Submits a task to compute values for the property sheet of an
+     * AbstractNode to a thread pool dedicated to such tasks with descriptively
+     * named threads (node-background-task-N).
+     *
+     * @param task The task.
+     *
+     * @return The Future of the task, may be used for task cancellation by
+     *         calling Future.cancel(true).
+     */
+    public static Future<?> submitTask(AbstractNodePropertySheetTask task) {
+        return executor.submit(task);
+    }
+
+    /**
+     * Constructs an abstract base class for background tasks needed to compute
+     * values for the property sheet of an AbstractNode.
+     *
+     * The results of the computation are returned by firing a
+     * PropertyChangeEvent and the run method has an exception firewall with
+     * logging. These features relieve the AbstractNode from having to create a
+     * thread to block on the get() method of the task Future.
+     *
+     * Only weak references to the AbstractNode and its PropertyChangeListener
+     * are held prior to task execution so that a queued task does not interfere
+     * with garbage collection if the node has been destroyed by the NetBeans
+     * framework.
+     *
+     * A thread pool with descriptively named threads (node-background-task-N)
+     * is provided for executing instances of the tasks.
+     *
+     * @param node     The node.
+     * @param listener A property change listener for the node.
+     */
+    protected AbstractNodePropertySheetTask(AbstractNode node, PropertyChangeListener listener) {
+        this.weakNodeRef = new WeakReference<>(node);
+        this.weakListenerRef = new WeakReference<>(listener);
+    }
+
+    /**
+     * Computes the values for the property sheet of an AbstractNode. The
+     * results of the computation are returned as a PropertyChangeEvent which is
+     * fired to the PropertyChangeEventListener of the node.
+     *
+     * IMPORTANT: Implementations of this method should check for cancellation
+     * by calling Thread.currentThread().isInterrupted() at approoraite
+     * intervals.
+     *
+     * @param node The AbstractNode.
+     *
+     * @return The result of the computation as a PropertyChangeEvent.
+     */
+    protected abstract PropertyChangeEvent computePropertyValue(AbstractNode node) throws Exception;
```

Comment:
Lets assume the background task logic we are submitting is for a LocalFileNode. Its routine depends entirely on the result of LocalFileNode.hasContentChildren(). How will I be able to utilize this class for my use case?

Reply:
Using generic, as suggested.